### ISSUE TITLE
Setting up ulsp on Neovim

### ISSUE THREAD

**Issue by IanTrudel** [uniconproject/unicon#457]
Anyone has got [ulsp](https://github.com/uniconproject/unicon/tree/master/uni/ulsp) working on Neovim? Neovim expects to connect to an LSP server through [vim.lsp.rpc.connect()](https://neovim.io/doc/user/lsp.html#vim.lsp.rpc.connect()) but ulsp is not listening on the port given. It seems that despite having a server.icn, it is actually a client.

```
./ulsp --socket 6010
open(:6010) ERROR: Connection refused
open(:6010) ERROR: Connection refused
open(:6010) ERROR: Connection refused
open(:6010) ERROR: Connection refused
```

**Comment by Jafaral** [comment]
The lsp server  was developed with an vscode extension,  and you are right, it does open the socket as a client back to the vscode extension.  It is easy to to add an option to run the socket in server mode.

Regardless who open the socket,  the lsp is still implementing the server side of the lsp protocol,  i.e, responding to the client requests.

**Comment by Jafaral** [comment]
@mstreeter10 , is it possible to change the lsp socket to run in server mode even with vscode? In all cases, we should add an option to open to open the socket in server mode to accept connections.

**Comment by IanTrudel** [comment]
Thank you @Jafaral. It should be mentioned that `make INSTALL` does not install ulsp either. Would be nice.

**Comment by Jafaral** [comment]
`ulsp` and `udap` are still in beta and we haven't added them to `make install`. `udap` was actually dropped in recent week in favor if `udb --adapter` option.  So things are still in flux. Until now, we assumed if you are using those you know what you are doing or at least you come and ask questions. If you can get any of these to work with `vim`, that would be  great. `emacs` and `sublime text` are on my radar  too.

**Comment by Jafaral** [comment]
BTW,  we also haven't tried running `ulsp` and `udap` out of tree, so expect failure there too.

**Comment by IanTrudel** [comment]
There are so many hidden gems in Unicon. A recent commit revealed the LSP. I wouldn't have known otherwise.

> BTW,  we also haven't tried running `ulsp` and `udap` out of tree, so expect failure there too. 

You may have noticed that I have a knack for finding failures. :)

> RE: vim

I will gladly share my Neovim configuration once it works.

**Comment by mstreeter10** [comment]
Ideally, the language server would run as a server by default. I'll look into this, vscode should be able to operate with the language server running as a server with a few tweaks to the extension.

**Comment by Jafaral** [comment]
adding `-s` as an option is also acceptable.  I kinda like the idea of being able to do it both ways so we cab support multiple use case. @mstreeter10 ,  we can do `ulsp -s --socket PORT`, and make that open `PORT` in listen mode, instead if connecting to `PORT`. Similar to how we do `udb --adapter` which I think run in server mode by default.

**Comment by mstreeter10** [comment]
Ideally, the language server would run as a server by default. I'll look into this, vscode should be able to operate with the language server running as a server with a few tweaks to the extension.

**Comment by IanTrudel** [comment]
> Ideally, the language server would run as a server by default. I'll look into this.

Take note that most language servers are dealing with stdin and stdout. This would be the easiest way to configure.

Godot is one of the rare LSPs to use a TCP connection, allowing an external editor to use it. All other LSPs I use are using stdin/stdout (C/C++/C#, Rust, Python, etc.)

Either way is fine with me but it's worth mentioning.

**Comment by mstreeter10** [comment]
This is now addressed in #468
